c10/core/Scalar.h

```diff
@@ -186,6 +195,27 @@ class C10_API Scalar {
   } v;
 };
 
+struct OptionalScalarRef {
+  OptionalScalarRef() : empty_(Scalar::MakeEmptyScalar{}), scalar_(empty_) {}
+  OptionalScalarRef(const Scalar& scalar) : scalar_(scalar) {}
+
+  bool has_value() const {
+    return scalar_.hasValue();
+  }
+
+  const Scalar& toScalar() const {
+    return scalar_;
+  }
+
+  operator bool() const {
+    return has_value();
+  }
+
+ private:
+  const Scalar empty_;
```

Comment:
The purpose of this member is so that `scalar_` has something to refer to when the `OptionalScalarRef` instance refers to nothing. I tried to make it `static` in order to ensure there is only one instance per class and not per object, but I ran into linker errors I couldn't fix. I put the definition of said `static` member into `Scalar.cpp` and I suspect it there are some objects that include `Scalar.h` that aren't linked against whatever library `Scalar.cpp` ends up in. I dug through CMakeLists/Bazel files for a while but couldn't figure out how to fix this.

A fix for the above or another way to create something for `scalar_` to refer to when the instance of `OptionalScalarRef` refers to nothing would be appreciated. 🙏

Reply:
Since your fields here are private, it seems better to just store a `Scalar*`. Then we don't even need `HAS_n`.